src/Microsoft.Extensions.Options/OptionsCache.cs

```diff
@@ -1,49 +1,65 @@
-// Copyright (c) .NET Foundation. All rights reserved.
+﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
-using System.Collections.Generic;
-using System.Threading;
+using System.Collections.Concurrent;
 
 namespace Microsoft.Extensions.Options
 {
-    internal class OptionsCache<TOptions> where TOptions : class, new()
+    /// <summary>
+    /// Used to cache TOptions instances.
+    /// </summary>
+    /// <typeparam name="TOptions">The type of options being requested.</typeparam>
+    public class OptionsCache<TOptions> : IOptionsCache<TOptions> where TOptions : class
     {
-        private readonly Func<TOptions> _createCache;
-        private object _cacheLock = new object();
-        private bool _cacheInitialized;
-        private TOptions _options;
-        private IEnumerable<IConfigureOptions<TOptions>> _setups;
+        private readonly ConcurrentDictionary<string, Lazy<TOptions>> _cache = new ConcurrentDictionary<string, Lazy<TOptions>>(StringComparer.Ordinal);
 
-        public OptionsCache(IEnumerable<IConfigureOptions<TOptions>> setups)
+        public virtual TOptions GetOrAdd(string name, Func<TOptions> createOptions)
         {
-            _setups = setups;
-            _createCache = CreateOptions;
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+            if (createOptions == null)
+            {
+                throw new ArgumentNullException(nameof(createOptions));
+            }
+            return _cache.GetOrAdd(name, new Lazy<TOptions>(createOptions)).Value;
```

Comment:
Sure so we are ok with the factory occasionally running more than once, since only one will 'win' and get stored.  That said, today we are using LazyInitialzer to guard against this I believe... https://github.com/aspnet/Options/blob/dev/src/Microsoft.Extensions.Options/OptionsCache.cs#L41